src/msg_recorder/replayer.cc

```diff
@@ -46,6 +50,10 @@ void MessageReplayer::MainLoop() {
             record_readers_.Push(std::move(top_reader));
         }
     }
+
+    if (!shutdown_flag_.load() && on_completion_) {
```

Comment:
如果是shutdown的，需不需要也做一个回调。 也许有场景可能会用到。

Reply:
shutdown就析构了，那析构之后手动做就好了

Reply:
@xkszltl shutdown flag indicates that we are stopping the main loop thread, so we should not do anything unnecessary but stop it.

Reply:
No, or you got one more "cancelled" state in state machine, and you'll need to distinguish between pre-exit and pre-cancel hook, and you need a mutex to block user from cancellation between this if and the execution of pre-exit hook.

Reply:
> No. I will just treated as a cancelation failure and the replayer will just exit as is.

Then you'll need a way to mark this failure somewhere, and you'll have confusing behavior that exit hook may or may not be executed after cancellation, and it will look exactly like what a missing mutex would lead to.
And there can still be needs for a post-cancellation hook, to do the clean up or summarization, so I would rather just have the share the same hook at this point.

Reply:
It is really a "failure", because we never have a "cancel" API. User can **stop** the replayer, and the replayer will always stop. Here we only distinguish the cases that whether the replaying is complete when it stopped.

Reply:
So I guess what we need are:

- On Canceled: called when the replaying does not complete
- On Complete: called when it completes
- On EXit: called after the callbacks above, no matter it completes or not.

@xkszltl wdyt?

Reply:
You don't need a separated "On Exit" because user can just reg twice to both. However if you have both "On Canceled" and "On Complete", it's possible that user cancel after complete, which will have ambiguity on whether to exec "On Canceled", that's why I proposed a single post-exit hook, regardless of whether it's cancelled or not.

If you want to have separated "On Canceled" and "On Complete" hooks, you should make it explicit on API that shutdown is just a signal, have a publicly visible state machine, and decide which hook to call purely base on that.

Reply:
Ack.